clang/lib/3C/Constraints.cpp

```diff
@@ -438,13 +434,32 @@ bool Constraints::graphBasedSolve() {
     if (!Res) {
       std::set<VarAtom *> Rest;
       Env.doCheckedSolve(true);
-      for (VarAtom *VA : Conflicts) {
-        assert(VA != nullptr);
-        std::string Rsn = "Bad pointer type solution";
-        Geq *ConflictConstraint = createGeq(VA, getWild(), Rsn);
+      for (auto Conflict : Conflicts) {
+        auto ConflictAtom = Conflict->getTargetNode().getData();
+        assert(ConflictAtom != nullptr);
+        ReasonLoc Rsn1 = Conflict->EdgeConstraint->getReason();
+        // determine a second reason
+        auto Succs = Conflict->getTargetNode().getEdges();
+        ReasonLoc Rsn2;
```

Comment:
This is only looking at immediate successors, which in general is not going to be sufficient. 

In this example, it correctly identifies that `&` is a problem, but it doesn't follow ptype constraints far enough to see that the `[0]` is also an issue.

```c
void foo() {
  int a;
  int *b = &a;
  int *c = b;
  c[0];
}
```

Reply:
That's correct, and I realize it was a known goal going into this task. But the prep work represented here is significant, and this PR is already 3 different parts. Lets save the complex graph search for the next PR. I don't mind doing it myself, but we should see how far this takes us and get a better sense of what additional info is needed.